### PR TITLE
#10851: support GQA in FlashDecode

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Ticket
#10851

### Problem description
Add support for GQA with Flash Decode

### What's changed
Reshaping input and output tensors to simulate GQA in FlashDecode MQA. 
batch_MQA = batch_GQA * num_k_heads_GQA
num_k_heads_MQA = 1
num_q_heads_MQA = num_q_heads_GQA/num_k_heads_GQA

### Checklist
- [x] Post commit CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/10321428064)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp [hunk 1]
```diff
@@ -449,6 +565,79 @@ Tensor scaled_dot_product_attention_decode(
...
+    auto k_shape = input_tensor_k.get_legacy_shape();
+    uint32_t B = k_shape[0];
+    uint32_t NQH = q_shape[1];
+    uint32_t NKH = k_shape[1];
+    uint32_t D = k_shape[3];
+
+    auto output_tensor = output_tensors.at(0);
+    output_tensor = ttnn::to_layout(output_tensor, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device*)nullptr);
+    output_tensor = ttnn::reshape(output_tensor, ttnn::Shape{std::array<uint32_t, 4>{1, B, NQH, D}});
+    output_tensor = ttnn::to_layout(output_tensor, ttnn::TILE_LAYOUT, std::nullopt, std::nullopt, (Device*)nullptr);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:638]:
```suggestion
    output_tensor = ttnn::to_layout(output_tensor, ttnn::ROW_MAJOR_LAYOUT);
    output_tensor = ttnn::reshape(output_tensor, ttnn::Shape{std::array<uint32_t, 4>{1, B, NQH, D}});
    output_tensor = ttnn::to_layout(output_tensor, ttnn::TILE_LAYOUT);
```

File: ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp [hunk 2]
```diff
@@ -449,6 +565,79 @@ Tensor scaled_dot_product_attention_decode(
...
+            uint32_t B = k_shape[0];
+            uint32_t NQH = q_shape[1];
+            uint32_t NKH = k_shape[1];
+            uint32_t D = k_shape[3];
+            uint32_t NG = NQH/NKH;
+
+            input_tensor_q = ttnn::to_layout(input_tensor_q, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device*)nullptr);
+            input_tensor_q = ttnn::transpose(input_tensor_q, 1, 2);
+            input_tensor_q = ttnn::reshape(input_tensor_q, ttnn::Shape{std::array<uint32_t, 4>{1,B*NKH, NG, D}});
+            input_tensor_q = ttnn::to_layout(input_tensor_q, ttnn::TILE_LAYOUT, std::nullopt, std::nullopt, (Device*)nullptr);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:602]:
```suggestion
            input_tensor_q = ttnn::to_layout(input_tensor_q, ttnn::ROW_MAJOR_LAYOUT);
            input_tensor_q = ttnn::transpose(input_tensor_q, 1, 2);
            input_tensor_q = ttnn::reshape(input_tensor_q, ttnn::Shape{std::array<uint32_t, 4>{1,B*NKH, NG, D}});
            input_tensor_q = ttnn::to_layout(input_tensor_q, ttnn::TILE_LAYOUT);
```

File: ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp [hunk 3]
```diff
@@ -351,6 +355,118 @@ const operation::Hash ScaledDotProductAttentionDecode::compute_program_hash(
...
+    return sdpa_decode_multi_core(
+        input_tensor_q,
+        input_tensor_k,
+        input_tensor_v,
+        output_tensor,
+        this->cur_pos,
+        scale,
+        this->compute_kernel_config,
+        this->program_config,
+        this->k_chunk_size);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:462]: You don't even use `q_chunk_size` or `k_chunk_size` here, did you intend to remove these and the `std::visit`?

File: ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp (hunk 1 above)
Comment (sraizada-tt) [reply ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:638]: This gives the error: 
```
../ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:599:45: note: in instantiation of function template specialization 'ttnn::decorators::operation_t<1, reflect::fixed_string<char, 16UL - 1>{{{116, 116, 110, 110, 58, 58, 116, 111, 95, 108, ...}}}, ttnn::operations::core::ToLayout, true>::operator()<tt::tt_metal::Tensor &, const tt::tt_metal::Layout &>' requested here
  599 |             input_tensor_q = ttnn::to_layout(input_tensor_q, ttnn::ROW_MAJOR_LAYOUT);
      |                                             ^
../ttnn/cpp/ttnn/operations/core/to_layout/to_layout_op.hpp:28:19: note: candidate function
   28 |     static Tensor operator()(
      |                   ^
../ttnn/cpp/ttnn/operations/core/to_layout/to_layout_op.hpp:35:19: note: candidate function
   35 |     static Tensor operator()(
```

File: ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp (hunk 2 above)
Comment (sraizada-tt) [reply ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:602]: This gives the error: 
```
../ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sdpa/sdpa_op.cpp:599:45: note: in instantiation of function template specialization 'ttnn::decorators::operation_t<1, reflect::fixed_string<char, 16UL - 1>{{{116, 116, 110, 110, 58, 58, 116, 111, 95, 108, ...}}}, ttnn::operations::core::ToLayout, true>::operator()<tt::tt_metal::Tensor &, const tt::tt_metal::Layout &>' requested here
  599 |             input_tensor_q = ttnn::to_layout(input_tensor_q, ttnn::ROW_MAJOR_LAYOUT);
      |                                             ^
../ttnn/cpp/ttnn/operations/core/to_layout/to_layout_op.hpp:28:19: note: candidate function
   28 |     static Tensor operator()(
      |                   ^
../ttnn/cpp/ttnn/operations/core/to_layout/to_layout_op.hpp:35:19: note: candidate function
   35 |     static Tensor operator()(
```

**Review by patrickroberts — APPROVED** [review]

**Pull request merged**